### PR TITLE
ENH: optimize: interpolative decomposition redundancy removal for linprog

### PR DESCRIPTION
**Pull request by mdhaber**
Supersedes gh-9783.

Adds a redundancy removal method based on the randomized interpolative decomposition to `linprog`. Also exposes redundancy removal method selection as an option.

In gh-9783, we discussed the possibility of redundancy removal using deterministic RRQR, which would not need the matrix rank a prior. As matrix rank calculation is much faster than these redundancy removal methods, we calculate the matrix rank first in order to determine whether redundancy removal is necessary. Consequently, the matrix rank is available for randomized interpolative decomposition, so this should be faster than the deterministic RRQR method. 

When gh-11245 is merged, I'll rebase this branch on master and add tests.

### CODE REVIEW

File: scipy/optimize/_linprog_util.py
Comment (Kai-Striega) [inline scipy/optimize/_linprog_util.py]: I don't see how these lines have any impact. Perhaps you missed adding ``_remove_redundancy_id(...)``?

File: scipy/optimize/_remove_redundancy.py
Comment (Kai-Striega) [inline scipy/optimize/_remove_redundancy.py]: One way would be to pass a tolerance value instead. ``interp_decomp`` can interoperate `k` in two possible ways; the required rank or the relative error (i.e. epsilon). The latter occurs if k < 1, in this case the signature becomes: 

``k, idx, proj = interp_decomp(A, eps_or_k)``

which would allow you to use `k` for the slicing later.

**Review by Kai-Striega — CHANGES_REQUESTED** [review]
The gist of this looks good to me. There are some questions/comments throughout, but these are minor. 

I can't see any tests yet which I think should be added before these are merged.

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [reply scipy/optimize/_linprog_util.py]: Maybe you were only looking at the first commit?

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: Thanks, I know about that use. Actually I meant to change that comment - I have to calculate the rank anyway to determine whether redundancy removal should be performed at all - as all of these routines are slower than np.linalg.matrix_rank. If I have it, I might as well use it!

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: 👍

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: k

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: 👍

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: 👍

**Comment by mdhaber** [comment]
> I can't see any tests yet which I think should be added before these are merged.

Mention in original post that I'm waiting on gh-11245, which restructures the tests for the redundancy removal routines. When that is merged, I'll rebase this branch on master and add tests.

**Comment by mdhaber** [comment]
@Kai-Striega I changed the function names as you suggested in gh-11245, and the redundancy removal tests are now performed on the new interpolative decomposition routine, too. 

Before this is merged, though, I still need to:

- [ ] Test calling `linprog` with this redundancy removal option
- [ ] Add the new interpolative decomposition-based method to the redundancy removal benchmarks

**Comment by tylerjereddy** [comment]
> Before this is merged, though, I still need to

@mdhaber Can you let me know if your TODO tasks are likely to happen in the next week--otherwise I'll bump to `1.6.0`?

**Comment by mdhaber** [comment]
@tylerjereddy bumped to 1.6.0

**Comment by mdhaber** [comment]
New method is faster for most benchmarks and more reliable:
```
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] · For scipy commit 714dfc5b <expose_rr_option>:
[  0.00%] ·· Benchmarking virtualenv-py3.7-Cython0.29.21-Tempita-numpy-pybind11-pytest-six
[ 25.00%] ··· Running (optimize_linprog.Netlib_RR.time_netlib_rr--).
[ 75.00%] ··· optimize_linprog.Netlib_RR.time_netlib_rr                                                         ok
[ 75.00%] ··· ================================================================== ========== ============
                                            method                                problems
              ------------------------------------------------------------------ ---------- ------------
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         AFIRO      483±80μs
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         BLEND     6.25±0.3ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         FINNIS     80.5±7ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         RECIPE     61.9±1ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         SCSD6      512±40ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        VTP-BASE    31.5±1ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         BORE3D     403±30ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         CYCLE     8.27±0.2s
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         DEGEN2     776±20ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         DEGEN3    7.23±0.3s
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        ETAMACRO    703±50ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        PILOTNOV   2.74±0.06s
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>          QAP8        n/a
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         RECIPE     61.9±1ms
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        SCORPION   4.70±0.2s
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         SHELL     3.11±0.2s
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         SIERRA    22.7±0.7s
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         WOOD1P    2.00±0.08s
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    AFIRO      13.9±1ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    BLEND      66.6±7ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    FINNIS     91.8±5ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    RECIPE     112±3ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    SCSD6      428±20ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   VTP-BASE    97.9±4ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    BORE3D       n/a
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    CYCLE     11.1±0.2s
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    DEGEN2     556±30ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    DEGEN3    3.48±0.3s
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   ETAMACRO    775±70ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   PILOTNOV      n/a
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>     QAP8     17.2±0.6s
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    RECIPE     112±3ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   SCORPION    740±30ms
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    SHELL     1.86±0.4s
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    SIERRA    2.23±0.1s
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    WOOD1P    1.39±0.06s
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     AFIRO     1.97±0.2ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     BLEND      9.58±2ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     FINNIS     18.6±2ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     RECIPE    13.0±0.6ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     SCSD6      70.9±1ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    VTP-BASE    11.5±1ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     BORE3D    72.9±10ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     CYCLE     12.7±0.4s
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     DEGEN2     357±20ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     DEGEN3    4.85±0.2s
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    ETAMACRO    381±20ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    PILOTNOV   2.98±0.1s
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>      QAP8     5.17±0.2s
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     RECIPE    13.0±0.6ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    SCORPION    261±20ms
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     SHELL     1.71±0.08s
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     SIERRA    2.44±0.1s
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     WOOD1P     286±40ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         AFIRO     973±100μs
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         BLEND     2.94±0.5ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         FINNIS     15.7±2ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         RECIPE    4.85±0.3ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SCSD6      154±20ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        VTP-BASE   5.28±0.7ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         BORE3D     108±10ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         CYCLE     8.83±0.3s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         DEGEN2     175±10ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         DEGEN3    2.45±0.1s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        ETAMACRO    282±7ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        PILOTNOV   1.93±0.2s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>          QAP8     3.78±0.1s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         RECIPE    4.85±0.3ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        SCORPION    166±6ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SHELL     1.19±0.04s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SIERRA    1.34±0.07s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         WOOD1P     399±30ms
              ================================================================== ========== ============

[100.00%] ··· optimize_linprog.Netlib_RR.track_netlib_rr                                                        ok
[100.00%] ··· ================================================================== ========== ========
                                            method                                problems
              ------------------------------------------------------------------ ---------- --------
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         AFIRO      0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         BLEND      0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         FINNIS     0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         RECIPE     0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         SCSD6      0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        VTP-BASE    0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         BORE3D     0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         CYCLE      0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         DEGEN2     0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         DEGEN3     0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        ETAMACRO    0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        PILOTNOV    0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>          QAP8      n/a
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         RECIPE     0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>        SCORPION    0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         SHELL      0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         SIERRA     0.0
                   <function _remove_redundancy_svd at 0x000001F88ED3D168>         WOOD1P     1.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    AFIRO      0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    BLEND      0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    FINNIS     0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    RECIPE     0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    SCSD6      0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   VTP-BASE    0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    BORE3D     n/a
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    CYCLE      0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    DEGEN2     0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    DEGEN3     0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   ETAMACRO    0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   PILOTNOV    n/a
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>     QAP8      0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    RECIPE     0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>   SCORPION    0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    SHELL      0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    SIERRA     0.0
               <function _remove_redundancy_pivot_sparse at 0x000001F88ED3D0D8>    WOOD1P    -129.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     AFIRO      0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     BLEND      0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     FINNIS     0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     RECIPE     0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     SCSD6      0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    VTP-BASE    0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     BORE3D     0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     CYCLE      0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     DEGEN2     0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     DEGEN3     0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    ETAMACRO    0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    PILOTNOV    0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>      QAP8      0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     RECIPE     0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>    SCORPION    0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     SHELL      0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     SIERRA     0.0
               <function _remove_redundancy_pivot_dense at 0x000001F88ED3D048>     WOOD1P     1.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         AFIRO      0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         BLEND      0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         FINNIS     0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         RECIPE     0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SCSD6      0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        VTP-BASE    0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         BORE3D     0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         CYCLE      0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         DEGEN2     0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         DEGEN3     0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        ETAMACRO    0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        PILOTNOV    0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>          QAP8      0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         RECIPE     0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        SCORPION    0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SHELL      0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SIERRA     0.0
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         WOOD1P     0.0
              ================================================================== ========== ========
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>        SCORPION    166±6ms
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SHELL     1.19±0.04s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         SIERRA    1.34±0.07s
                    <function _remove_redundancy_id at 0x000001F88ED3D1F8>         WOOD1P     399±30ms
              ================================================================== ========== ============
```

File: benchmarks/benchmarks/optimize_linprog.py
Comment (mdhaber) [inline benchmarks/benchmarks/optimize_linprog.py:21]: PEP8

File: benchmarks/benchmarks/optimize_linprog.py
Comment (mdhaber) [inline benchmarks/benchmarks/optimize_linprog.py:72]: PEP8 here and below.

**Comment by mdhaber** [comment]
Hey @Kai-Striega I think this is ready.

**Comment by tylerjereddy** [comment]
@mdhaber @Kai-Striega Is this close to ready or will we need to bump the milestone?

**Comment by Kai-Striega** [comment]
I think this looks ready (and like a good improvement), unless @mdhaber has any further changes he'd like to add in?

**Comment by mdhaber** [comment]
No changes; I'll just resolve merge conflicts.

**Comment by mdhaber** [comment]
Just ran the benchmarks after merging changes; still looks good.

File: scipy/optimize/_remove_redundancy.py
Comment (tylerjereddy) [inline scipy/optimize/_remove_redundancy.py:509]: "a linear combinations" -> singular? I don't want to flush CI for a comment typo fix at this point though

File: scipy/optimize/tests/test__remove_redundancy.py
Comment (tylerjereddy) [inline scipy/optimize/tests/test__remove_redundancy.py:160]: Adding the old-school `assert_` is probably not desirable anymore now that `pytest` allows us to use `assert` and sometimes even provides better feedback from its "overload" of that functionality

**Review by tylerjereddy — APPROVED** [review]
CI is green apart from known issue with NumPy pre-release wheels and I'm mostly trusting the approving expert/core dev review here.

I only have two minor review comments from reading through the changes just now, that are not sufficient to justify another round of revisions/CI flushes at this time, so I'll merge it in.

**Comment by tylerjereddy** [comment]
Thanks @mdhaber @Kai-Striega

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py:509]: Thanks Tyler. Added these to linprog cleanup list in gh-9671

**Pull request merged**
